crates/ecs/src/lib.rs

```diff
@@ -503,74 +503,355 @@ pub struct World {
+            return Err(WorldError::ComponentTypeAlreadyExistsForEntity(
+                entity,
+                TypeId::of::<ComponentType>(),
+            ));
+        }
+
+        let target_archetype_exists = self.get_archetype_with_matching_components(&target_type_ids);
+
+        Ok((target_archetype_exists, (source_type_ids, target_type_ids)))
+    }
+
+    fn move_entity_components_between_archetypes(
+        &mut self,
+        entity: Entity,
+        target_archetype_idx: ArchetypeIndex,
+        components_to_move: Vec<TypeId>,
     ) -> WorldResult<()> {
-        self.store_entity_in_archetype(entity, 0)?;
+        let source_archetype_index = *self
+            .entity_to_archetype_index
+            .get(&entity)
+            .ok_or(WorldError::EntityDoesNotExist(entity))?;
 
-        let big_archetype = self
+        let source_archetype = self
             .archetypes
-            .get_mut(0)
-            .ok_or(WorldError::ArchetypeDoesNotExist(0))?;
+            .get(source_archetype_index)
+            .ok_or(WorldError::ArchetypeDoesNotExist(source_archetype_index))?;
+
+        let source_component_vec_idx = *self
+            .archetypes
+            .get(source_archetype_index)
+            .ok_or(WorldError::ArchetypeDoesNotExist(source_archetype_index))?.entity_to_component_index
+            .get(&entity)
+            .expect("Entity should have a component index tied to it since it is already established to be existing within the archetype");
+
+        // move_element removes component from source but does not move if no ComponentVec
+        // with matching type exists in target
+        for type_id in components_to_move {
+            source_archetype.component_typeid_to_component_vec
+                .get(&type_id)
+                .expect("Type that tried to be fetched should exist
+                     in archetype since types are fetched from this archetype originally.")
+                .move_element(source_component_vec_idx,
+                              self.archetypes
+                                  .get(target_archetype_idx)
+                                  .expect("Archetype should exist since it was the requirement for entering this matching case"))
+                .map_err(WorldError::CouldNotMoveComponent)?;
+        }
+
+        let target_archetype: &mut Archetype = self
+            .archetypes
+            .get_mut(target_archetype_idx)
+            .expect(
+            "Archetype should exist since it was the requirement for entering this matching case",
+        );
+
+        target_archetype
+            .entity_to_component_index
+            .insert(entity, target_archetype.entity_to_component_index.len());
+
+        target_archetype.entity_order.push(entity);
+        self.entity_to_archetype_index
+            .insert(entity, target_archetype_idx);
+
+        Ok(())
+    }
+
+    fn move_entity_components_to_new_archetype(
```

Comment:
Yeah this seems feasible! Will implement if I can't find any issues

Reply:
I'm using move_element_to_new when moving to a new archetype which creates a component_vec according to the generic fetched from the component_vec I'm calling. This is the only way I found to extract and use a components type during move. Since I'm creating a new component vector in the new archetype this requires the newly created archetype to be mutable (since I'm updating its structure). This mutable requirement is not necessary for an already existing archetype since it already contains the vectors for each component type that is moved.

A solution where we move elements to unassigned component vectors and update them all and assign them to a new archetypes could perhaps work? It seems to be alot of work restructuring so I will avoid doing this atm.

Reply:
While executing `move_entity_components_to_new_archetype` we need to fetch the component vec of all relevant type from the source archetype. The component vec is then used to perform the the execution of `move_element_to_new`. This function requires the source archetype (by extension of the fetch operation previously) to be borrowed from world. 

In addition to this,  `move_element_to_new` requires a target archetype which is mutable in order for us to add a component vector of the same type as the one being used from the source archetype and to update its data structures to reflect the addition of this new component vector. 

The result (from my understanding) is that we cannot borrow both the source and target archetypes at once from world since target needs to be mutable. My way of getting around this is by having world not having ownership over the target archetype until we are done constructing it and all elements have been moved. 

The crux is that I cannot create new component vectors of relevant types in target archetype since haven't found a way to pass types from the already existing component vectors in the source archetype to the target archetype without being present in the specific component_vec. As such I needed to create the component vecs for the target archetype within the component vec of the source archetype. 

If I were to add the target archetype to world before sending its index to  `move_entity_components_between_archetypes` then it would not be possible to generate the required component vectors since the borrow rules would be violated.

If a way to pass types is found we could probably return this from the source archetypes component vec, drop the borrow and then mutably add a new component vec of this  type to the target archetype. 

I found it strange that this was the only solution I found so there may be that my knowledge is lacking and that this could be done much easier and in a more structured way.

Is it clearer now or is there something that I´m missing in my explanation/ reasoning?

Reply:
I see. While it's a bit difficult to get two mutable references to different indices in an array, it's entirely possible. I've implemented a version using that and pushed it to the branch in [this commit](https://github.com/martinjonsson01/recs/commit/6aba05181fe8763039c03ba3403a5da3d411e381). Take a look, in case I made any assumptions that breaks your code (but the tests are passing at least)!

Reply:
Thanks, I had no idea about split_at_mut. Great solution to the issue!